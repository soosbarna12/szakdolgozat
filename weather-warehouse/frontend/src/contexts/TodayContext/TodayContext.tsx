import React, { createContext, useEffect, useState } from "react";
import { TodayContextType, TodayLocationData, TodayProviderProps } from "./TodayContext.type";


export const TodayContext = createContext<TodayContextType>({
	location: { name: "", lat: 0, lon: 0 },
	setLocation: () => { },
});

export const TodayLocationProvider = ({ children }: TodayProviderProps) => {

	const [location, setLocation] = useState<TodayLocationData>(() => {
		const storedLocation = localStorage.getItem("todayLocation");
		const defaultLocation = { name: "", lat: 0, lon: 0 };
		try {
			return storedLocation ? JSON.parse(storedLocation) : defaultLocation;
		} catch (error) {
			return defaultLocation;
		}
	});

	useEffect(() => {
		localStorage.setItem("todayLocation", JSON.stringify(location));
	}, [location]);


	return (
		<TodayContext.Provider value={{ location, setLocation }}>
			{children}
		</TodayContext.Provider>
	);
};
