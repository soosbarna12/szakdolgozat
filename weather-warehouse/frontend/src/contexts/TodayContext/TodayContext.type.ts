import { Dispatch, ReactNode, SetStateAction } from "react";

// location name interfere
export interface TodayLocationData {
  name: string;
  country?: string;
  state?: string;
  lat: number;
  lon: number;
}

export interface TodayContextType {
  location: TodayLocationData;
  setLocation: Dispatch<SetStateAction<TodayLocationData>>;
}

export interface TodayProviderProps {
	children: ReactNode;
}
