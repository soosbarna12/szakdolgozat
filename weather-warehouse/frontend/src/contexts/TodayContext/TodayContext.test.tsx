import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import React, { useContext } from 'react';
import { TodayContext, TodayLocationProvider } from './TodayContext';

describe('contexts/TodayContext', () => {
  const mockChildren = <div data-testid="mockChildren">Mock Children</div>;

  beforeEach(() => {
    jest.clearAllMocks();
    localStorage.clear();
  });

  it('matches TodayLocationProvider component snapshot', () => {
    const { container } = render(
      <TodayLocationProvider>{mockChildren}</TodayLocationProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('provides default values when localStorage is empty', () => {
    render(
      <TodayLocationProvider>{mockChildren}</TodayLocationProvider>
    );
  
    const defaultLocation = { name: '', lat: 0, lon: 0 };
    expect(localStorage.getItem('todayLocation')).toEqual(
      JSON.stringify(defaultLocation)
    );
  
    const context = screen.getByTestId('mockChildren').closest('div');
    expect(context).toBeInTheDocument();
  });

  it('loads values from localStorage when available', () => {
    const mockLocation = { name: 'New York', lat: 40.7128, lon: -74.006 };
    localStorage.setItem('todayLocation', JSON.stringify(mockLocation));

    render(
      <TodayLocationProvider>{mockChildren}</TodayLocationProvider>
    );

    expect(localStorage.getItem('todayLocation')).toEqual(JSON.stringify(mockLocation));
  });

  it('updates localStorage when location changes', () => {
    const TestComponent = () => {
      const { location, setLocation } = useContext(TodayContext);

      return (
        <div>
          <button
            data-testid="updateLocationButton"
            onClick={() => setLocation({ name: 'Los Angeles', lat: 34.0522, lon: -118.2437 })}
          >
            Update Location
          </button>
          <span data-testid="locationName">{location.name}</span>
        </div>
      );
    };

    render(
      <TodayLocationProvider>
        <TestComponent />
      </TodayLocationProvider>
    );

    const updateButton = screen.getByTestId('updateLocationButton');
    fireEvent.click(updateButton);

    expect(localStorage.getItem('todayLocation')).toEqual(
      JSON.stringify({ name: 'Los Angeles', lat: 34.0522, lon: -118.2437 })
    );
    expect(screen.getByTestId('locationName').textContent).toBe('Los Angeles');
  });

  it('handles invalid JSON in localStorage gracefully', () => {
    localStorage.setItem('todayLocation', 'invalid-json');
  
    render(
      <TodayLocationProvider>{mockChildren}</TodayLocationProvider>
    );
  
    const defaultLocation = { name: '', lat: 0, lon: 0 };
    expect(localStorage.getItem('todayLocation')).toEqual(
      JSON.stringify(defaultLocation)
    );
  });
});