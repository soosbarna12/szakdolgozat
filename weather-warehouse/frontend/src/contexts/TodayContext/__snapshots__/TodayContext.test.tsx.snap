// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contexts/TodayContext matches TodayLocationProvider component snapshot 1`] = `
<div>
  <div
    data-testid="mockChildren"
  >
    Mock Children
  </div>
</div>
`;
